Linux UI improvements

robxnano opened the pull request:
Here are some tweaks and updates for the GTK UI. I hope you'll find them useful enough to merge.
Changes:
- Add keyboard shortcut support by replacing the application menu with a GMenu (closes #4107, closes #4306)
- Save space in in the source open dialog by putting all the options on a single line
- Improve the appearance of the settings dialog
- Add automatic dark theme switching on DEs that support freedesktop.org color-scheme property
- Make the main toolbar icons symbolic so they change color on dark themes (fixes #2879)
- Added context menu in the presets dialog
- Colored warning buttons for message dialogs

If you have any questions or problems, let me know.

**Test on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [x] Ubuntu Linux (22.04)

**Screenshots:**
![Main Window](https://user-images.githubusercontent.com/89391914/199365267-6f487ac9-e31d-49ea-9968-e09eaafedcf0.png)
Keyboard shortcuts:
![Menus](https://user-images.githubusercontent.com/89391914/199365688-f3d281be-513d-4c9a-8583-207d8bc21988.png)

robxnano committed:
Improvements to GTK source open dialog

- Used set_extra_widget to add options
- Aligned options horizontally to save space
- Added title
- Added suggested-action button style

robxnano committed:
Added XDG color-scheme implementation

Handbrake will now react to changes in the desktop environment
color scheme and change theme accordingly.

robxnano committed:
LinGUI: Improve layout of preferences dialog

- Replace OK button with close button to make it clearer that
  changes are saved immediately
- Replace GtkNotebook with GtkStack
- Move StackSwitcher into title bar to save space

robxnano committed:
LinGUI: Replace old menu widgets with GMenuModel

- The new model is easier to maintain and makes it easy to add
  keyboard shortcuts.
- Added keyboard shortcuts based on the Windows version.

robxnano committed:
LinGUI: Add warning color to dialog buttons

The Yes/OK button of a message dialog will be red (to warn
the user that the action is destructive) if the flag
GTK_MESSAGE_WARNING is passed to it.
- Also added new function to create message dialogs with
  a title and subtitle. This will need translation work
  before it can be used though.

robxnano committed:
Added warning colors to more message dialogs

robxnano committed:
LinGUI: Make main toolbar icons symbolic

The old toolbar icons didn't show up well with dark themes.
Now they now change color automatically based on the theme.
Also changed a few other icons to the freedesktop standard versions.

robxnano committed:
LinGUI: Use context menu in presets list

Removed the menu bar from the presets dialog and replaced it with
a context menu in order to make it easier to delete presets.

robxnano committed:
LinGUI: Bug fixes

robxnano committed:
LinGUI: Fix Wayland window icon

Use g_set_prgname to set the program name so that KWin
can find the desktop file and set the window icon.
Fixes https://github.com/HandBrake/HandBrake/issues/4182

sr55 commented:
@jstebbins  if you have time to take a quick look, that would be good.

I'll try make some time at the weekend to do some testing.    Thank you for the submission @robxnano

robxnano committed:
LinGUI: More icon improvements

- Renamed the icons so they show up inverted in more places
- Replaced the stop icon as it was pixelated at 2X scale
- Recreated the play and pause icons in SVG format

robxnano committed:
LinGUI: Refactor file chooser filter code

robxnano committed:
Removed calls to unused widgets

robxnano committed:
LinGUI: Use GtkActionBar for status bar

- Window no longer jumps around when the progress bar appears
- Added When Done combo box to status bar like Windows version
(fixes https://github.com/HandBrake/HandBrake/issues/1513)
- Replaced title menu with better looking version

galad87 commented:
I can't comment on the gtk code changes, but I tested it a bit and it seems to be ok.
I would leave the menu bar in the presets window too. The alt+key shortcuts don't work on my Ubuntu VM, they open the menu that starts with the pressed letter instead.

robxnano commented:
Thanks for the feedback! I missed that problem with the shortcuts. But I'm not sure if it would be better to change the menu accelerators rather than choose different shortcuts. I just copied most of the shortcuts from the Windows version, but maybe there should be a discussion about which ones would be the most appropriate.

Regarding the menu bar in the presets dialog, there were several reasons why I chose to remove it:
Having a menu bar in a dialog is unusual, most applications only have menu bars in the main window.
With only one toplevel item, it doesn't obviously look like a menu. It could be a window title like several other dialogs have.
The GtkMenu used to render it is deprecated and removed in GTK4. The GMenu API which I used for the main window's menu bar only allows the menu bar to be set per-application, so if HandBrake does get ported to GTK4, it wouldn't be possible to keep it.

Of course, I'm happy to restore it. But if you just want a more obvious way of accessing the context menu, I would suggest replacing it with a standalone menu button, like the ones used in the Queue dialog. That would solve all the issues of discoverability and future-proofing.

robxnano committed:
LinGUI: New D-Bus calls for shutdown and suspend

The old org.freedesktop.PowerManagement interface was removed,
which meant that shutdown and suspend didn't work on many
distributions. The new equivalent uses org.freedesktop.login1.
To make it work in Flatpak, the --system-talk-name permission
is required.
Fixes https://github.com/HandBrake/HandBrake/issues/4002

robxnano committed:
LinGUI: Remove org.gnome.SessionManager D-Bus call

This interface is Gnome-specific, and no longer needed for
shutdown as org.freedesktop.login1 handles it along with suspend.

galad87 commented:
Yes it needs a obvious way to access the context menu, it's just a thing that would be nice to have imho.

robxnano committed:
LinGUI: Don't exit maximized mode on crop change

Fixes https://github.com/HandBrake/HandBrake/issues/4224

sr55 commented:
With regards to the presets, the Windows has a small toolbar within the preset overlay. Exposes the most common functions.  Maybe a way of doing it.

Otherwise, I've also given it a test pass.

I'm not sure @jstebbins  has any free time at the minute.  If not, we can accept and if needed any tidyup later.


P.s, we are kicking around on IRC on the libera-chat network.  #handbrake and #handbrake-dev if you ever feel like it

robxnano committed:
LinGUI: Added Actions button to Presets window

- Changed shortcuts to not clash with keyboard accelerators
- Removed unnecessary -symbolic suffix from icon names
  (improves compatibility with some themes and icon sets)
- Various other UI tweaks

robxnano commented:
I added a toolbar with a dropdown menu to the presets window. Does this look okay?
![image](https://user-images.githubusercontent.com/89391914/200596131-d88049e3-4181-4228-9f37-57f39e8e7d21.png)

Although I don't have much time at the moment to do any more, I'll try to drop in on the chat some time soon.

galad87 commented:
The set default menu item is always disabled in the main window menu and in the presets actions menu, but it's enable in the contextual menu when right clicking on a preset.

Anyway, it looks good! A lot of nice fixes, I would merge it if no one has any complain.

robxnano committed:
Fix action names so that set default works

sr55 commented:
Thanks @robxnano 

On the offchance you get some free time / want to contribute more,  https://github.com/HandBrake/HandBrake/issues?q=is%3Aissue+is%3Aopen+label%3ALinuxUI  May be of interest :)

I generally tag Linux UI specific issues with the LinuxUI tag.
